src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs

```diff
@@ -230,15 +230,25 @@ private static SafeAccessTokenHandle DuplicateAccessToken(IntPtr accessToken)
 
         private static SafeAccessTokenHandle DuplicateAccessToken(SafeAccessTokenHandle accessToken)
         {
-            Debug.Assert(accessToken != null);
-
             if (accessToken.IsInvalid)
             {
                 return accessToken;
             }
 
-            SafeAccessTokenHandle duplicateAccessToken = DuplicateAccessToken(accessToken.DangerousGetHandle());
-            GC.KeepAlive(accessToken);
+            SafeAccessTokenHandle duplicateAccessToken = SafeAccessTokenHandle.InvalidHandle;
+            IntPtr currentProcessHandle = Interop.Kernel32.GetCurrentProcess();
+            if (!Interop.Kernel32.DuplicateHandle(
+                    currentProcessHandle,
+                    accessToken,
+                    currentProcessHandle,
+                    ref duplicateAccessToken,
+                    0,
+                    true,
+                    Interop.DuplicateHandleOptions.DUPLICATE_SAME_ACCESS))
+            {
+                throw new SecurityException(new Win32Exception().Message);
+            }
```

Comment:
Rather than declaring the DllImport twice, you could also just do the AddRef/Release yourself:
```C#
SafeAccessTokenHandle duplicateAccessToken;
bool success = false;
try
{
    accessToken.DangerousAddRef(ref success);
    duplicateAccessToken = DuplicateAccessToken(accessToken.DangerousGetHandle());
}
finally
{
    if (success) accessToken.DangerousRelease();
}
return duplicateAccessToken;
```

Reply:
What's the purpose of the bool return? It seems like it always either returns true or throws.

Reply:
Same as the ref bool argument to Monitor.Enter; reliability on platforms with thread aborts.

Reply:
What if success is false after DangerousAddRef? Should it throw ObjectDisposedException? Maybe I can just completely ignore the return value.

Reply:
Ah I see

Reply:
> What if success is false after DangerousAddRef?

It won't be: as you said, it'll either be true or DangerousAddRef will throw.  The ref bool is there to support thread aborts.  On systems with thread aborts, you put the DangerousAddRef call inside the try, and then if there's an exception, the finally can look at the bool to determine whether it needs to DangerousRelease.  If the DangerousAddRef were outside the try, a thread abort could occur after a successful AddRef but before entering the try, and the SafeHandle would be leaked.  

> Maybe I can just completely ignore the return value.

On .NET Core, you can.